homeassistant/components/axis.py

```diff
@@ -0,0 +1,347 @@
+            {'id': CONF_INCLUDE,
+             'name': 'Device functionality (space separated list)',
+             'type': 'text'},
+            {'id': ATTR_LOCATION,
+             'name': 'Physical location of device (optional)',
+             'type': 'text'},
+            {'id': CONF_SCAN_INTERVAL,
+             'name': 'Sensor update interval (optional)',
+             'type': 'number'},
+        ]
+    )
+
+
+def setup(hass, base_config):
+    """Common setup for Axis devices."""
+    configurator = get_component('configurator')
+    configurator._get_instance(hass)  # pylint: disable=W0212
+
+    def _shutdown(call):  # pylint: disable=unused-argument
+        """Stop the metadatastream on shutdown."""
+        for name, device in AXIS_DEVICES.items():
+            _LOGGER.info("Stopping metadatastream for %s.", name)
+            device.stop_metadatastream()
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, _shutdown)
+
+    def axis_device_discovered(service, discovery_info):
+        """Called when axis devices has been found."""
+        host = discovery_info['host']
+        name = discovery_info['hostname']
+        serialnumber = discovery_info['properties']['macaddress']
+
+        if serialnumber not in AXIS_DEVICES:
+            config_file = _read_config(hass)
+            if serialnumber in config_file:
+                config = config_file[serialnumber]
+                config['hass'] = hass
+                if not setup_device(config):
+                    _LOGGER.error('Couldn\'t set up %s', config['name'])
+            else:
+                request_configuration(hass, name, host, serialnumber)
+
+    discovery.listen(hass, 'axis_mdns', axis_device_discovered)
+
+    if DOMAIN in base_config:
+        for device in base_config[DOMAIN]:
+            config = base_config[DOMAIN][device]
+            if CONF_NAME not in config:
+                config[CONF_NAME] = device
+            config['hass'] = hass
+            if not setup_device(config):
+                _LOGGER.error('Couldn\'t set up %s', config['name'])
+
+    return True
+
+
+def setup_device(config):
+    """Set up device."""
+    from axis import AxisDevice
+
+    hass = config['hass']
+    device = AxisDevice(config)  # Initialize device
+    enable_metadatastream = False
+
+    if device.serial_number is None:
+        # If there is no serial number a connection could not be made
+        _LOGGER.error('Couldn\'t connect to %s', config[CONF_HOST])
+        return False
+
+    for component in config[CONF_INCLUDE]:
+        if component in EVENT_TYPES:
+            # Sensors are created by device calling event_initialized
+            # when receiving initialize messages on metadatastream
+            device.add_event_topic(convert(component, 'type', 'subscribe'))
+            if not enable_metadatastream:
+                enable_metadatastream = True
+        else:
+            discovery.load_platform(hass, component, DOMAIN, config)
```

Comment:
It doesn't have a 'hard coupling' to hass. The Device doesn't know up front how many sensors it will register with hass so I set a callback to the device that it will know to execute when a new event is registered. That function is event_initialized. This also allows new events to be added dynamically if configuration changes on the device.

This probably is a bit convoluted but it has worked great so far.